internal/admin/admin.go

```diff
@@ -172,13 +175,24 @@ func (adm *Admin) Close() (err error) {
 func (adm *Admin) Metadata(ctx context.Context) (*varlogpb.MetadataDescriptor, error) {
 	adm.mu.RLock()
 	defer adm.mu.RUnlock()
+
 	return adm.mrmgr.ClusterMetadataView().ClusterMetadata(ctx)
 }
 
 func (adm *Admin) getStorageNode(ctx context.Context, snid types.StorageNodeID) (*admpb.StorageNodeMetadata, error) {
 	adm.mu.RLock()
 	defer adm.mu.RUnlock()
 
+	snm, err, _ := adm.sfg.Do(sfgkey.GetStorageNodeKey(snid), func() (interface{}, error) {
+		return adm.getStorageNodeInternal(ctx, snid)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return snm.(*admpb.StorageNodeMetadata), nil
+}
+
+func (adm *Admin) getStorageNodeInternal(ctx context.Context, snid types.StorageNodeID) (*admpb.StorageNodeMetadata, error) {
```

Comment:
How about using singleflight once more inside `ClusterMetadata`?

Reply:
Good idea. In next followup PR, I will remove unnecessary locks in internal/admin/mrmanager.(*mrManager) and use singleflight.